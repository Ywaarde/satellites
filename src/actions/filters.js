// SET TEXT FILTER
export const setTextFilter = (text = '') => ({
  type: 'SET_TEXT_FILTER',
  text
});

// SORT BY MISSION ID
export const sortByMissionID = () => ({
  type: 'SORT_BY_MISSION_ID'
});

// SORT BY MISSION SHORT NAME
export const sortByMissionShortName = () => ({
  type: 'SORT_BY_MISSION_SHORT_NAME'
});

// SORT BY MISSION AGENCIES
export const sortByMissionAgencies = () => ({
  type: 'SORT_BY_MISSION_AGENCIES'
});

// SORT BY MISSION STATUS
export const sortByMissionStatus = () => ({
  type: 'SORT_BY_MISSION_STATUS'
});

// SET CURRENT PAGE
export const setCurrentPage = (pageNo) => ({
  type: 'SET_PAGE',
  pageNo
});