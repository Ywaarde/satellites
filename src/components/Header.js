import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <div>
      <h1>Satellites</h1>
      <NavLink to='/' activeClassName='is-active' exact={true}>Misson Overview</NavLink>
      <NavLink to='/create' activeClassName='is-active'>Add mission</NavLink>
    </div>
)

export default Header;