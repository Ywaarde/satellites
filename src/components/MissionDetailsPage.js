import React from 'react';

const MissionDetailsPage = (props) => (
    <div>
    This is the page with details for Mission ID: {props.match.params.id}.
    </div>
);

export default MissionDetailsPage;