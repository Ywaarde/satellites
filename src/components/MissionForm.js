import React from 'react';

export default class MissionForm extends React.Component {
  constructor(props){
    super(props);

    this.state = {
      MissionID: props.mission ? props.mission.mission_id : props.new_id,
      MissionNameShort: props.mission ? props.mission.name_short : '',
      MissionNameFull: props.mission ? props.mission.name_full : '',
      MissionAgencies: props.mission ? props.mission.agencies : '',
      MissionStatus: props.mission ? props.mission.status : ''
    };
  }

  onShortNameChange = (e) => {
    const MissionNameShort = e.target.value;
    this.setState({ MissionNameShort });
  }

  onFullNameChange = (e) => {
    const MissionNameFull = e.target.value;
    this.setState({ MissionNameFull });
  }

  onAgenciesChange = (e) => {
    const MissionAgencies = e.target.value;
    this.setState({ MissionAgencies });
  }

  onMissionStatusChange = (e) => {
    const MissionStatus = e.target.value;
    this.setState({ MissionStatus });
  }

  onSubmit = (e) => {
    e.preventDefault();
    this.props.onSubmit(this.state);
  }

  render() {
    return (
      <div>
        {this.props.mission && <h3>Mission ID {this.props.mission.mission_id}</h3>}
        {this.props.new_id && <h3>Mission ID {this.props.new_id}</h3>}
        <form onSubmit={this.onSubmit}>
          <input
            type='text'
            placeholder='Mission Short Name'
            required={true}
            autoFocus
            value={this.state.MissionNameShort}
            onChange={this.onShortNameChange}
          />
          <textarea
            placeholder='Mission Full Name'
            required={true}
            value={this.state.MissionNameFull}
            onChange={this.onFullNameChange}
          >
          </textarea>
          <input
            type='text'
            placeholder='Mission Agencies'
            required={true}
            value={this.state.MissionAgencies}
            onChange={this.onAgenciesChange}
          />
          <input
            type='text'
            placeholder='Mission Status'
            required={true}
            value={this.state.MissionStatus}
            onChange={this.onMissionStatusChange}
          />
          <button>Save Mission</button>
        </form>
      </div>
    );
  }
}

// Full mission details
    // Mission ID*
    // Mission Name Short*
    // Mission Name Full*
    // Mission Agencies*
    // Mission MissionStatus*
    // Launch Date,
    // EOL Date,
    // Applications,
    // Instruments,
    // Instrument IDs,
    // Orbit Type,
    // Orbit Altitude,
    // Orbit Period,
    // Orbit Inclination,
    // Repeat Cycle,
    // Orbit LST,
    // Orbit Longitude,
    // Orbit Sense,
    // Mission Site,
    // Data Access Portal
