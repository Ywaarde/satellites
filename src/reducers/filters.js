// Filter reducer

const filterReducerDefaultState = {
    text: '',
    sortBy: 'mission_id',
    page: 1,
    perPage: 50
  };
  
  const filterReducer = (state = filterReducerDefaultState, action) => {
    switch (action.type) {
      case 'SET_TEXT_FILTER':
        return {
          ...state,
          text: action.text
        };
      case 'SORT_BY_MISSION_ID':
        return {
          ...state,
          sortBy: 'mission_id'
        };
      case 'SORT_BY_MISSION_SHORT_NAME':
        return {
          ...state,
          sortBy: 'short_name'
        };
      case 'SORT_BY_MISSION_AGENCIES':
        return {
          ...state,
          sortBy: 'agencies'
        };
      case 'SORT_BY_MISSION_STATUS':
        return {
          ...state,
          sortBy: 'status'
        };
      case 'SET_PAGE':
        return {
          ...state,
          page: action.pageNo
        };
      default:
        return state;
    }
  };

  export default filterReducer;