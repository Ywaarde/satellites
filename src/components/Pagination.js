import React from 'react';
import { connect } from 'react-redux';
import { setCurrentPage } from '../actions/filters';

export class Pagination extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      totalMissions: '',
      totalPages: '',
      currentPage: '',
      pages: ''
    };
  }

  componentDidMount() {
    this.setState({
      totalMissions: this.props.missions.length,
      totalPages: Math.ceil(this.props.missions.length / this.props.filters.perPage),
      currentPage: this.props.filters.page
    });
  }

  setPage(page) {
    if (page < 1) {
      page = 1;
    } else if (page > this.state.totalPages) {
      page = this.state.totalPages;
    }

    this.setState({
      currentPage: page
    });

    this.props.onChangePage(page);
  }

  getPager() {
    let pages = []

    for (let i = 1; i <= this.state.totalPages; i++) {
      pages.push(i);
    }
    return {
      pages
    };
  }

  render() {
    var pager = this.getPager();

    return (
      <p className="pagination">
        <button
          disabled={this.state.currentPage === 1}
          onClick={() => this.setPage(1)}
        >
          First
          </button>
        <button
          disabled={this.state.currentPage === 1}
          onClick={() => this.setPage(this.state.currentPage - 1)}
        >
          Previous
          </button>
        {pager.pages.map((page) => (
          <button
            key={page}
            className={this.state.currentPage === page ? "active" : ""}
            onClick={() => this.setPage(page)}
          >
            {page}
          </button>
        ))}
        <button
          disabled={this.state.currentPage === this.state.totalPages}
          onClick={() => this.setPage(this.state.currentPage + 1)}
        >
          Next
          </button>

        <button
          disabled={this.state.currentPage === this.state.totalPages}
          onClick={() => this.setPage(this.state.totalPages)}
        >
          Last
          </button>
      </p>
    );
  }
}

const mapStateToProps = (state, props) => ({
  missions: state.missions,
  filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
  onChangePage: (page) => dispatch(setCurrentPage(page))
});

export default connect(mapStateToProps, mapDispatchToProps)(Pagination)