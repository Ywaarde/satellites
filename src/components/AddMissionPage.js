import React from 'react';
import { connect } from 'react-redux';
import MissionForm from './MissionForm';
import { addMission } from '../actions/missions';

export class AddMissionPage extends React.Component {
  onSubmit = (mission) => {
    this.props.addMission(mission);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <h3>Add New Mission</h3>
        <MissionForm
          new_id = {this.props.new_id}
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  const highest_id = state.missions.reduce((prev, current) => (prev.mission_id > current.mission_id) ? prev : current);
  return {
      new_id: highest_id.mission_id + 1
  };
};

const mapDispatchToProps = (dispatch) => ({
  addMission: (mission) => dispatch(addMission(mission))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddMissionPage);