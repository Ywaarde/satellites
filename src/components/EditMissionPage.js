import React from 'react';
import { connect } from 'react-redux';
import MissionForm from './MissionForm';
import { editMission, removeMission } from '../actions/missions';

export class EditMissionPage extends React.Component {
  onSubmit = (mission) => {
    this.props.editMission(this.props.mission.id, mission);
    this.props.history.push('/');
  }
  onClick = () => {
    this.props.removeMission(this.props.mission.id);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <MissionForm
          mission = {this.props.mission}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onClick}
        >Delete Mission</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  mission: state.missions.find((mission) => mission.mission_id.toString() === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editMission: (id, mission) => dispatch(editMission(id, mission)),
  removeMission: (id) => dispatch(removeMission({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditMissionPage);