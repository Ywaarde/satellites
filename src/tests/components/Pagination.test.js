import React from 'react'
import { shallow } from 'enzyme'
import { Pagination } from '../../components/Pagination'
import testMissions from '../fixtures/missions'
import {altFilters} from '../fixtures/filters'

let onChangePage, wrapper

beforeEach(() => {
    onChangePage = jest.fn()
    wrapper = shallow(
        <Pagination 
            onChangePage={onChangePage} 
            missions={testMissions} 
            filters={altFilters}/>)
})

test('should render pagination correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle page change', () => {
    wrapper.find('button').at(3).simulate('click')
    expect(onChangePage).toHaveBeenLastCalledWith(2)
})