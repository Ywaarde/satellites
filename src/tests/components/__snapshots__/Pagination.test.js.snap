// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render pagination correctly 1`] = `
<p
  className="pagination"
>
  <button
    disabled={true}
    onClick={[Function]}
  >
    First
  </button>
  <button
    disabled={true}
    onClick={[Function]}
  >
    Previous
  </button>
  <button
    className="active"
    key="1"
    onClick={[Function]}
  >
    1
  </button>
  <button
    className=""
    key="2"
    onClick={[Function]}
  >
    2
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    Next
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    Last
  </button>
</p>
`;
